core/src/main/java/org/apache/iceberg/deletes/RoaringPositionBitmap.java

```diff
@@ -0,0 +1,309 @@
+import org.roaringbitmap.RoaringBitmap;
+
+/**
+ * A bitmap that supports positive 64-bit positions, but is optimized for cases where most positions
+ * fit in 32 bits by using an array of 32-bit Roaring bitmaps. The bitmap array is grown as needed
+ * to accommodate the largest value.
+ *
+ * <p>Incoming 64-bit positions are divided into a 32-bit "key" using the most significant 4 bytes
+ * and a 32-bit position using the least significant 4 bytes. For each key in the set of positions,
+ * a 32-bit Roaring bitmap is maintained to store a set of 32-bit positions for that key.
+ *
+ * <p>To test whether a certain position is set, its most significant 4 bytes (the key) are used to
+ * find a 32-bit bitmap and the least significant 4 bytes are tested for inclusion in the bitmap. If
+ * a bitmap is not found for the key, then the position is not set.
+ *
+ * <p>Note that positions must be greater than or equal to 0 add less than {@link #MAX_POSITION}.
+ * This bitmap does not support positions larger than {@link #MAX_POSITION} as the bitmap array
+ * length is a signed 32-bit integer that must be greater than or equal to 0.
+ */
+class RoaringPositionBitmap {
+
+  static final long MAX_POSITION = toPosition(Integer.MAX_VALUE - 1, Integer.MIN_VALUE);
+  private static final RoaringBitmap[] EMPTY_BITMAP_ARRAY = new RoaringBitmap[] {};
+  private static final long BITMAP_COUNT_SIZE_BYTES = 8L;
+  private static final long BITMAP_KEY_SIZE_BYTES = 4L;
+
+  private RoaringBitmap[] bitmaps;
+
+  RoaringPositionBitmap() {
+    this.bitmaps = EMPTY_BITMAP_ARRAY;
+  }
+
+  private RoaringPositionBitmap(RoaringBitmap[] bitmaps) {
+    this.bitmaps = bitmaps;
+  }
+
+  /**
+   * Adds a position to the bitmap.
+   *
+   * @param pos the row position
+   */
+  public void add(long pos) {
+    validatePosition(pos);
+    int high = highBytes(pos);
+    int low = lowBytes(pos);
+    allocateBitmapsIfNeeded(high + 1 /* required number of bitmaps */);
+    bitmaps[high].add(low);
+  }
+
+  /**
+   * Adds a range of positions to the bitmap.
+   *
+   * @param posStartInclusive the start position of the range (inclusive)
+   * @param posEndExclusive the end position of the range (exclusive)
+   */
+  public void addRange(long posStartInclusive, long posEndExclusive) {
+    for (long pos = posStartInclusive; pos < posEndExclusive; pos++) {
+      add(pos);
+    }
+  }
+
+  /**
+   * Checks if the bitmap contains a position.
+   *
+   * @param pos the row position
+   * @return true if the position is in this bitmap, false otherwise
+   */
+  public boolean contains(long pos) {
+    validatePosition(pos);
+    int high = highBytes(pos);
+    int low = lowBytes(pos);
+    return high < bitmaps.length && bitmaps[high].contains(low);
+  }
+
+  /**
+   * Adds all positions from the other bitmap to this bitmap, modifying this bitmap in place.
+   *
+   * @param that the other bitmap
+   */
+  public void or(RoaringPositionBitmap that) {
```

Comment:
Why not return this for chaining?

Reply:
To highlight it modifies the bitmap in place, not a strong opinion. What do you think?

Reply:
Either way is ok with me, I know in most of our public APIs we use the functional approach even for modify in place.